### ISSUE TITLE
API endpoint to the profiles API that can grab the URL from a GitRepository

### ISSUE THREAD

**Issue by Himangini** [weaveworks/profiles#316]
### User story 2:
**As a profile installer**, who would like to add a private profile to my cluster, I can:
- Apply the provided `Secret` to my cluster
- Run `pctl add`
- See that the correct files are generated \o/

**Details:**
This is the part that Profiles is providing, you can **assume that there is a GitRepository already existing with the private profiles**.

We'll also need to add an API endpoint to the profiles API that can grab the URL from a `GitRepository` that contains a `Profile` and return it to the client (`pctl`).

This API endpoint, should accept a request for a URL for a Profile, we should calculate which `GitRepository` contains the profile, and the path within the repository, and return something like...

```json
{
  "archiveURL": "https://flux-git-repository.archive/tarball.tgz", 
  "path": "profiles/path-to-profile"
}
```
The _archiveURL_ comes from the Status of the Flux GitRepository

We'll update the client (`pctl`) to make a call to do this, and instead of cloning the source, we fetch from the tarball and use the cloned source in the profiles directory. (_It may make sense to do this for always, and not have a switch based on public/private._)

**Comment by aclevername** [comment]
## The problem

Talking through @bigkevmcd I'm not sure if we want to play this story yet. I think there was 1 assumption made during planning that actually makes this whole thing a bit more complicated

> This is the part that Profiles is providing, you can assume that there is a GitRepository already existing with the private profiles.

This would mean that we have to have a pre-existing `GitRepository` resource per profile name & tag. In practise how would this work? Do we scan a repository and create a `GitRepository` for every single tag and leave it around permanently? This would use up a lot of storage and github API calls very quickly. We do a similar thing when scanning a repository for profiles but this is short-lived and hopefully infrequent https://github.com/weaveworks/profiles/blob/e8d045e156a651b1046b889aadc0af85fe2e0050/pkg/gitrepository/manager.go#L51-L68

We went through a couple ideas of how this could work here https://miro.com/app/board/o9J_lyO3kAU=/ but all of the solution ultimately end up with a lot of complexity getting added.

TLDR: Moving to an API approach atm introduces a lot more complexity on how the `GitRepository` resources are created/managed. We might need to instead of doing this approach re-evaluate our approach to cloning as a whole

## User Value

What value was this story delivering? I wasn't involved in the original chat here https://github.com/weaveworks/profiles/issues/307 but it sounds like we shifted from "this doesn't scale so lets fix it" to "the UX isn't great, lets adjust it". I think (and feel free to correct me) the part about the UX we didn't like was providing company X with 1 deploy key, and them having to distribute this to every individual who needs to run `pctl add`. Having an API do the cloning for you remove this, as all the user needs to have access to to run `pctl add` is the catalog API instead, much simpler, but pushing a lot of complexity onto us. 

A solution to unblock this in the short-term if customers complain that distribution of the deploy key is unfeasible/painful we would instead ask them to maintain there own private mirror repository of our private repository. Its then just up to them to distribute access to that mirror repository using whatever conventions they already have (e.g. Adding github users to the repository to grant access). This isn't an "amazing" solution, but it at least provides a workaround that doesn't involve lots of complexity getting added on our part.

**Comment by aclevername** [comment]
@Callisto13 @Skarlso you both mentioned at standup you were interested in this ^

Please let me know if you disagree with anything from the above at all :smile:

**Comment by Callisto13** [comment]
I am not in the least interested 😉  but I can read an have an opinion I suppose

**Comment by Callisto13** [comment]
All sounds correct to me, and you are right that the mirroring thing is not amazing or pleasant at all. I can't see paying customers putting up with that for long. As a stop-gap it is "fine" I guess, but do we have a long-term plan here? Is the answer tied up in the other integration initiative we have going on?

**Comment by aclevername** [comment]
> All sounds correct to me, and you are right that the mirroring thing is not amazing or pleasant at all. I can't see paying customers putting up with that for long.

Is it *that* unpleasent? Everybody already setups up automation to ensure they have the most up to date software, normaly we are talking binaries/releases, here its just git repository. Your right its not great, but I don't really see it being a  massive pain for a customer


> As a stop-gap it is "fine" I guess, but do we have a long-term plan here? Is the answer tied up in the other integration initiative we have going on?

I think the answer is yes and no. We don't have a long-term plan, because having one now would probably be reasonably pointless until we get this in the hands of users and get feedback. We might shift in a very different direction very quickly. But I do think the other integration initiative will provide an alternative, which might be more palatable.

**Comment by Skarlso** [comment]
I would take the complexity on us. The mirror thing is a pain, yes. And I agree with Claudia that if I was a customer I wouldn't want to deal with that hassle at all, even if automated. :/

**Comment by aclevername** [comment]
> I would take the complexity on us. The mirror thing is a pain, yes. And I agree with Claudia that if I was a customer I wouldn't want to deal with that hassle at all, even if automated. :/

I'd disagree. I think I'd rather not have us take on all this extra complexity to preempt a problem we think that will come up before it does, when a relatively small workaround unblocks a customer in the short-term. Its a trade off for sure but I basically don't think we should spend too much time solving a problem before we undertand it better.

And then saying all that as said in my previous comment this might all not be a problem (at least in the same way) once we start the integration work as things are going to change :joy: :shrug:

**Comment by Skarlso** [comment]
Alright. Let's make this a 20:20 hindsight. 😂

**Comment by Callisto13** [comment]
> I think I'd rather not have us take on all this extra complexity to preempt a problem we think that will come up before it does

oh i don't disagree, and i am not saying let's not do it, but also let's be ready for some customers to be like "ew what?"

**Comment by aclevername** [comment]
> > I think I'd rather not have us take on all this extra complexity to preempt a problem we think that will come up before it does
> 
> oh i don't disagree, and i am not saying let's not do it, but also let's be ready for some customers to be like "ew what?"

:100: , its all trade offs at the end of the day

**Comment by Himangini** [comment]
@aclevername will add a new issue summarising the discussion from planning
We will pick this up after wego cli integration

**Comment by aclevername** [comment]
New issue, highlighting the problem we want to fix https://github.com/weaveworks/pctl/issues/320
